lib/network_manager.go

```diff
@@ -0,0 +1,787 @@
+		id, err := nm.CreateNonValidatorPersistentOutboundConnection(connectIp)
+		if err != nil {
+			glog.Errorf("NetworkManager.initiatePersistentConnections: Problem connecting "+
+				"to connectIp %v: %v", connectIp, err)
+			continue
+		}
+
+		nm.persistentIpToRemoteNodeIdsMap.Set(connectIp, id)
+	}
+}
+
+// ###########################
+// ## Validator Connections
+// ###########################
+
+// SetActiveValidatorsMap is called by the owner of the NetworkManager to update the activeValidatorsMap. This should
+// generally be done whenever the active validators set changes.
+func (nm *NetworkManager) SetActiveValidatorsMap(activeValidatorsMap *collections.ConcurrentMap[bls.SerializedPublicKey, consensus.Validator]) {
+	nm.activeValidatorsMapLock.Lock()
+	defer nm.activeValidatorsMapLock.Unlock()
+	nm.activeValidatorsMap = activeValidatorsMap.Clone()
+
+}
+
+func (nm *NetworkManager) getActiveValidatorsMap() *collections.ConcurrentMap[bls.SerializedPublicKey, consensus.Validator] {
+	nm.activeValidatorsMapLock.RLock()
+	defer nm.activeValidatorsMapLock.RUnlock()
+	return nm.activeValidatorsMap.Clone()
+}
+
+// refreshValidatorIndex re-indexes validators based on the activeValidatorsMap. It is called periodically by the
+// validator connector.
+func (nm *NetworkManager) refreshValidatorIndex(activeValidatorsMap *collections.ConcurrentMap[bls.SerializedPublicKey, consensus.Validator]) {
+	// De-index inactive validators. We skip any checks regarding RemoteNodes connection status, nor do we verify whether
+	// de-indexing the validator would result in an excess number of outbound/inbound connections. Any excess connections
+	// will be cleaned up by the NonValidator connector.
+	validatorRemoteNodeMap := nm.rnManager.GetValidatorIndex().ToMap()
+	for pk, rn := range validatorRemoteNodeMap {
+		// If the validator is no longer active, de-index it.
+		if _, ok := activeValidatorsMap.Get(pk); !ok {
+			nm.rnManager.SetNonValidator(rn)
+			nm.rnManager.UnsetValidator(rn)
+		}
+	}
+
+	// Look for validators in our existing outbound / inbound connections.
+	allNonValidators := nm.rnManager.GetAllNonValidators()
+	for _, rn := range allNonValidators {
+		// It is possible for a RemoteNode to be in the non-validator indices, and still have a public key. This can happen
+		// if the RemoteNode advertised support for the SFValidator service flag during handshake, and provided us
+		// with a public key, and a corresponding proof of possession signature.
+		pk := rn.GetValidatorPublicKey()
+		if pk == nil {
+			continue
+		}
+		// It is possible that through unlikely concurrence, and malevolence, two non-validators happen to have the same
+		// public key, which goes undetected during handshake. To prevent this from affecting the indexing of the validator
+		// set, we check that the non-validator's public key is not already present in the validator index.
+		if _, ok := nm.rnManager.GetValidatorIndex().Get(pk.Serialize()); ok {
```

Comment:
This will actually be common. It looks like every validator creates an outbound connection to every other validator, and so you will frequently have one outbount connectoin AND one inbound connectoin for each validator. Disconnecting here seems like it will screw up the outbound validator connections for all peers.